go/pkg/cas/upload.go

```diff
@@ -662,6 +677,160 @@ func (u *uploader) uploadBatch(ctx context.Context, reqs []*repb.BatchUpdateBlob
+		}
+
+		// Compress using an in-memory pipe. This is mostly to accomodate the fact
+		// that zstd package expects a writer.
+		// Note that using io.Pipe() means we buffer only bytes that were not uploaded yet.
+		pr, pw := io.Pipe()
+		enc.Reset(pw)
+
+		// Read from disk and make RPCs concurrently.
+		eg, ctx := errgroup.WithContext(ctx)
+		eg.Go(func() error {
+			switch _, err := enc.ReadFrom(r); {
+			case err == io.ErrClosedPipe:
+				// The other goroutine exited before we finished encoding.
+				// Might be a cache hit or context cancelation.
+				// In any case, the other goroutine has the actual error, so return nil
+				// here.
+				return nil
+			case err != nil:
+				return errors.Wrapf(err, "failed to read the file/blob")
+			}
+
+			if err := enc.Close(); err != nil {
+				return errors.Wrapf(err, "failed to close the zstd encoder")
+			}
+			return pw.Close()
+		})
+		eg.Go(func() error {
+			defer pr.Close()
+			return u.streamFromReader(ctx, pr, item.Digest, true, updateCacheStats, streamBuf)
+		})
+		return eg.Wait()
+	})
+}
+
+func (u *uploader) streamFromReader(ctx context.Context, r io.Reader, digest *repb.Digest, compressed, updateCacheStats bool, buf []byte) error {
+	stream, err := u.byteStream.Write(ctx)
+	if err != nil {
+		return err
+	}
+
+	req := &bspb.WriteRequest{}
+	if compressed {
+		req.ResourceName = fmt.Sprintf("%s/uploads/%s/compressed-blobs/zstd/%s/%d", u.InstanceName, uuid.New(), digest.Hash, digest.SizeBytes)
+	} else {
+		req.ResourceName = fmt.Sprintf("%s/uploads/%s/blobs/%s/%d", u.InstanceName, uuid.New(), digest.Hash, digest.SizeBytes)
+	}
+
+chunkLoop:
+	for {
+		// Before reading, check if the context if canceled.
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		// Read the next chunk from the pipe.
+		// Use ReadFull to ensure we aren't sending tiny blobs over RPC.
+		n, err := io.ReadFull(r, buf)
+		switch {
+		case err == io.EOF || err == io.ErrUnexpectedEOF:
+			req.FinishWrite = true
+		case err != nil:
+			return err
+		}
+		req.Data = buf[:n] // must limit by `:n` in ErrUnexpectedEOF case
+
+		// Send the chunk.
+		switch err = stream.Send(req); {
+		case err == io.EOF:
+			// The server closed the stream.
+			// Most likely the file is already uploaded, see the CommittedSize check below.
+			break chunkLoop
+		case err != nil:
+			return err
+		case req.FinishWrite:
+			break chunkLoop
+		}
+
+		// Prepare the next request.
+		req.ResourceName = "" // send the resource name only in the first request
+		req.WriteOffset += int64(len(req.Data))
+	}
+
+	// Finalize the request.
+	switch res, err := stream.CloseAndRecv(); {
+	case err != nil:
+		return err
+	case res.CommittedSize != digest.SizeBytes:
+		return fmt.Errorf("unexpected commitSize: got %d, want %d", res.CommittedSize, digest.SizeBytes)
+	}
+
+	// Update stats.
+	cacheHit := !req.FinishWrite
+	if !cacheHit {
+		atomic.AddInt64(&u.stats.Streamed.Bytes, digest.SizeBytes)
```

Comment:
_github won't let me reply on the original thread_

I think adding a "RealMoved" or "NetworkBytes" to `DigestStats`, so that we have more granularity between batch and stream, is better. It gives more granularity on the impact of compression, particularly when we get BatchCompression on the API (it'll happen at some point, just not in the near future). For now, `NetworkBytes` will be the same to `Bytes` for batches, not accounting for retries.

Reply:
added a todo